### PR TITLE
Add a Client Stack Module to rewrite HTTP Host Header/Consul integration for it

### PR DESCRIPTION
**Pull request by Ashald**
This PR introduces following features:
1) Consul Catalog Namer now has an extra option `domain` - when set it makes it to embed `l5d-http-host` meta field with value `$name.service.$dc.$domain` into `Addr.Bound`
2) Namerd Thrift Interface/Client now can pass around `Addr.Metadata` for `Addr.Bound` and `Addr.Inet` but only those with values of `String` type
3) `clientStack` within `io.buoyant.linkerd.protocol.HttpInitializer#defaultRouter` now includes `RewriteHostHeader` module that rewrites outgoing request's `Host` header but only if `l5d-http-host` meta field is present within `Addr.Bound` metadata

All changes are covered with tests.

### CODE REVIEW

**Comment by Ashald** [comment]
@olix0r please take a look at this PR whenever you'll have time. I'm interested in your 'conceptual' comments along with code review as I feel that my code if far away from being pretty. :/

**Comment by olix0r** [comment]
@Ashald thanks for your patience waiting for us to get to this.

My initial understanding of this feature was that the Host value you'd like to set was actually stored in Consul.  I now see that this is provided to the namer configuration statically in the namerd/linkerd configuration.   So, effectively, all names resolved through this namer must receive the same domain value.  If that's the case, why does need any integration at all with the consul namer?  Couldn't we just add a module to linkerd that sets the host value?  It seems odd to me that this is set globally on the namer rather than on a per-service basis.

**Comment by Ashald** [comment]
Hi @olix0r, thanks for taking your time to look a this one.

The short answer that configured value in namer is not a static thing that will be set for all requests but rather a required suffix to generate Consul-ish host value. I tired to explain that briefly within `linkerd/docs/namer.md`.

When Consul is used as a DNS server it resolves names that end with pre-configured TLD. By default it's `.consul` but can be changed to anything, for instance, `acme.co`. Given that TLD Consul will resolve following name 'patterns':
1) `$name.service.$dc.$tld` - some service (e.g., `foobar.service.us-east.acme.co`)
2) `$node.node.$dc.$tld` - some machine (e.g., `web-worker1.node.us-east.acme.co`)

As I said above, you can configure TLD in Consul config but unfortunately it doesn't return it when you query its HTTP API for service info. It only uses it to distinguish between names it's supposed to handle and others that it cannot handle.
We're heavily relying on HTTP `Host` header in order to distinguish between different services/versions of service hosted on single `ip:port`. I believe other Consul users might have similar use-cases.

The idea behind this PR is to allow to set the aforementioned TLD in Consul namer and use it to set `Host` header in the form `$name.service.$dc.$tld` whenever `linkerd` forwards request based on Consul concrete name.

Having this feature and services `foo-1` and `foo-2` we will be able to switch a version by just setting a dtab like `/foo=>/foo-1` and we won't need to generate additional web server configs to handle a 'canonical' name like `foo.service.us-east.acme.co`. Moreover it enables us to do an a/b or canary testing with different versions sitting on the same `ip:port`. I think it won't be useful only to us but is the crucial feature for many Consul users who rely on `Host` value.

**Comment by Ashald** [comment]
@olix0r upon further digging Consul API I found an endpoint that reports the domain value we need. I think it will make a lot of sense to integrate with it so we can fetch the domain from Consul directly. Given that the option in `Namer` will become a flag. The question is how to name it: `override_host`? Or something else?

**Comment by olix0r** [comment]
@Ashald thanks for the background (and sorry for missing your other explanation in the docs ;)

I agree that it would be preferable to fetch this value from consul directly, rather than have to synchronize configuration across consul and linkerd.

So:
- the consul namer propagates a domain suffix
- linkerd has some sort of stripDomain option that causes this value to be used to modify downstream requests

**Comment by Ashald** [comment]
@olix0r thanks for the review. So right now I have it implemented following way:
1) if given `l5d-http-host` meta field on bound addr `linekrd` will always set that as a host
2) there is an option on `ConsulCatalogNamer` whether to set that meta field or not

I see the benefit of such implementation in (1) that other components may use this feature as well (2) and you can enabled/disable that on per-namer basis.

Although what you mentioned above is supposed to work in an 'opposite direction' I don't mind changing the PR to make it work as you suggested. Though I'll need your advice on how to add that parameter to `linkerd`.

BTW, are you OK with meta field called `l5d-http-host`?

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala [hunk 1]
```diff
@@ -2,16 +2,20 @@ package io.buoyant.consul
...
-  includeTag: Boolean = false
+  catalogApi: v1.CatalogApi,
+  agentApi: v1.AgentApi,
+  includeTag: Boolean = false,
+  setHost: Boolean = false
 ) extends Namer {
 
   import CatalogNamer._
 
+  lazy val domainFuture: Future[String] = agentApi.self().map(_.Config.get.Domain.get.stripPrefix(".").stripSuffix("."))
```
Comment (olix0r) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: please try to wrap lines around 100 chars.  In this case I'd just move `agentApi...` onto the following line.

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala [hunk 2]
```diff
@@ -89,49 +93,42 @@ class CatalogNamer(
...
-
-    def mkRequest(): Future[Seq[ServiceNode]] =
-      api
-        .serviceNodes(key.name, datacenter = Some(datacenter), tag = key.tag, blockingIndex = Some(index), retry = true)
-        .map { indexedNodes =>
-          indexedNodes.index.foreach(setIndex)
-          indexedNodes.value
-        }
+    def mkRequest(index: Option[String]) =
+      catalogApi.serviceNodes(key.name, datacenter = Some(datacenter), tag = key.tag, blockingIndex = index, retry = true)
```
Comment (olix0r) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: similarly, I'd break this up like:

```scala
catalogApi.serviceNodes(
  key.name,
  datacenter = ...,
  tag = ...,
  blockingIndex = index,
  retry = true
)
```

File: namerd/iface/interpreter-thrift-idl/src/main/thrift/namer.thrift [hunk 3]
```diff
@@ -81,6 +81,7 @@ struct AddrReq {
 }
 
 struct AddrMeta {
+  1: map<string, string> fields
```
Comment (olix0r) [inline namerd/iface/interpreter-thrift-idl/src/main/thrift/namer.thrift]: I'd prefer to enumerate the set of supported metadata, until we really need an arbitrary map.

Let's have an optional `authority` field (to match HTTP/2 conventions) that captures what we're doing here.

File: consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package io.buoyant.consul.v1
+
+import com.twitter.conversions.time._
+import com.twitter.finagle.http
+import com.twitter.finagle.service.Backoff
+import com.twitter.finagle.stats.{DefaultStatsReceiver, StatsReceiver}
+import com.twitter.util.{Closable, Duration, Future}
+
+object AgentApi {
+  def apply(c: Client): AgentApi = new AgentApi(c, s"/$versionString")
```
Comment (olix0r) [inline consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala]: Where does _versionString_ come from

File: consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+class AgentApi(
+  override val client: Client,
+  override val uriPrefix: String,
+  override val backoffs: Stream[Duration] = Backoff.exponentialJittered(1.milliseconds, 5.seconds),
+  override val stats: StatsReceiver = DefaultStatsReceiver
```
Comment (olix0r) [inline consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala]: Omit `override`?  I don't think these actually override anything...

File: consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  // https://www.consul.io/docs/agent/http/agent.html#agent_self
+  def self(retry: Boolean = false): Future[Self] = {
+    val req = mkreq(http.Method.Get, s"$agentPrefix/self")
+    executeJson[Self](req, retry).map(_.value)
+  }
+}
+
+case class Config(Domain: Option[String])
+
+case class Self(Config: Option[Config])
```
Comment (olix0r) [inline consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala]: please lowercase field names (`domain`, `config`).

please include a ~1-sentence summary comment of each of these types so that a maintainer can get their bearings without having to read consul docs

**Comment by olix0r** [comment]
Thanks again for all the work on this.

Right now I think the main blocker is that I'd like to change is to move away from an arbitrary map of attributes to something typed.  Since the read side of this is a part of the core linkerd http functionality, I think it makes sense to introduce the metadata as a first-class primitive.

File: consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala (hunk 6 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala]: Regarding case of field names - here I followed example from https://github.com/BuoyantIO/linkerd/blob/master/consul/src/main/scala/io/buoyant/consul/v1/CatalogApi.scala#L62 - aren't field names should match JSON field names for parser to work properly? Or it can figure out the thing? Because in Consul JSON output they're starting from upper case.

File: consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala (hunk 4 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala]: The same as for other APIs (I inherited it from the original implementation) - from `consul/src/main/scala/io/buoyant/consul/v1/package.scala` - it defines `package object v1` with common stuff.

**Comment by Ashald** [comment]
@olix0r I pushed some changes and updated the PR as per your comments. Can you please take another look?

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala [hunk 7]
```diff
@@ -83,55 +93,58 @@ class CatalogNamer(
...
+      )
 
     def clear(): Unit = synchronized {
       updateableAddr() = Addr.Pending
     }
 
-    def init(): Future[Unit] = mkRequest().flatMap(update).handle(handleUnexpected)
+    def init(): Future[Unit] = domainFuture.flatMap { domain =>
+      val meta = if (setHost)
+        Addr.Metadata(("authority", s"${key.name}.service.$datacenter.$domain"))
```
Comment (olix0r) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: I'd prefer if we had one val'd version of this string so that typos were compilation errors.

Also minor style preference for `Addr.Metadata("authority" -> s"...")`.  Neither are blockers, just sharing...

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala [hunk 8]
```diff
@@ -83,55 +93,58 @@ class CatalogNamer(
...
+      else
+        Addr.Metadata.empty
+      mkRequest(None).flatMap(nodes => update(nodes, meta)).handle(handleUnexpected)
```
Comment (olix0r) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: tioli: it would be fine to write this as `mkRequest(None).flatMap(update(_, meta)).handle(...)`

Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: That's neat - thanks!

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala [hunk 9]
```diff
@@ -2,16 +2,21 @@ package io.buoyant.consul
...
+  catalogApi: v1.CatalogApi,
+  agentApi: v1.AgentApi,
+  includeTag: Boolean = false,
+  setHost: Boolean = false
 ) extends Namer {
 
   import CatalogNamer._
 
+  lazy val domainFuture: Future[String] =
+    agentApi.self().map(_.Config.get.Domain.get.stripPrefix(".").stripSuffix("."))
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: avoid `Option.get`.  instead do something like:
```
agentApi.self().flatMap { self =>
  val domain = self.Config
    .flatMap(_.Domain)
    .map(_.stripPrefix(".").stripSuffix("."))
  domain match {
    case Some(d) => Future.value(d)
    case None => Future.exception(???)
  }
}
```

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala (hunk 7 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: You mean, within this file? Or across the whole `linkerd` codebase?

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala [hunk 10]
```diff
@@ -83,55 +93,58 @@ class CatalogNamer(
...
-    }
-
-    def mkRequest(): Future[Seq[ServiceNode]] =
-      api
-        .serviceNodes(key.name, datacenter = Some(datacenter), tag = key.tag, blockingIndex = Some(index), retry = true)
-        .map { indexedNodes =>
-          indexedNodes.index.foreach(setIndex)
-          indexedNodes.value
-        }
+    def mkRequest(index: Option[String]) =
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: is this `index` refactor unrelated to the rest of the change?  If so, I'd prefer it split into a separate PR.

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala (hunk 7 above)
Comment (olix0r) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Ideally, it would be great to have the known metadata keys somewhere (as we do for headers, for instance) -- we can be incremental to getting it somewhere global, though.

File: consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    val req = mkreq(http.Method.Get, s"$agentPrefix/self")
+    executeJson[Self](req, retry).map(_.value)
+  }
+}
+
+// Classes below are used for JSON parsing hence field names case have to match the on in JSON
+
+// Represents configuration and member information of the local agent
+// For now only Config is parsed as other fields are unused
+case class Self(
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala]: What do you think about calling this `LocalAgent` or something like that?  `Self` (and `self` above) is a bit confusing because of scala's self-types.

**Comment by olix0r** [comment]
modulo nits and whatever alex has, this is looking about ready to me

File: namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala [hunk 12]
```diff
@@ -119,6 +120,10 @@ object HttpControlService {
     val prefix = s"$apiPrefix/addr"
   }
 
+  object ResolveUri extends NsPathUri {
```
Comment (adleong) [inline namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala]: I think this Resolve stuff snuck in from a different PR

File: namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala [hunk 13]
```diff
@@ -162,17 +173,16 @@ class ThriftNamerClient(
...
-          case Return(thrift.Addr(stamp1, thrift.AddrVal.Bound(thrift.BoundAddr(taddrs, _)))) =>
+          case Return(thrift.Addr(stamp1, thrift.AddrVal.Bound(thrift.BoundAddr(taddrs, boundMeta)))) =>
             val addrs = taddrs.map { taddr =>
-              val thrift.TransportAddress(ipbb, port, _) = taddr
+              val thrift.TransportAddress(ipbb, port, addressMeta) = taddr
               val ipBytes = Buf.ByteArray.Owned.extract(Buf.ByteBuffer.Owned(ipbb))
               val ip = InetAddress.getByAddress(ipBytes)
-              // TODO convert metadata
-              Address(new InetSocketAddress(ip, port))
+              Address.Inet(new InetSocketAddress(ip, port), convertMeta(addressMeta)).asInstanceOf[Address]
```
Comment (adleong) [inline namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala]: I don't think you need this asInstanceOf cast

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala (hunk 7 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Woul it be OK if I'll put it into an object in consul namespace for starters?

File: consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala (hunk 11 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala]: Sure

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala (hunk 9 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Thanks!

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala (hunk 10 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Yeah, a bit of it. I just saw a way to get rid of variable and changed the code a bit. But can put this into a separate PR.

File: namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala (hunk 12 above)
Comment (Ashald) [inline namerd/iface/control-http/src/main/scala/io/buoyant/namerd/iface/HttpControlService.scala]: Ooops, you right - I'll drop changes these from this PR

File: namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala (hunk 13 above)
Comment (Ashald) [inline namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala]: It didn't compile without it. :( Is there is a better way around?

Comment (adleong) [inline namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala]: Address.Inet is a subclass of Address so something weird must be going on here...

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala (hunk 9 above)
Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: According to Consul API spec the `Domain` field cannot be empty and thus `Future.exception(???)` should never be executed. Would be OK to leave it as is or I should replace `???` with some concrete exception (and which one would be appropriate in such case)?

File: namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala (hunk 13 above)
Comment (Ashald) [inline namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala]:
```
[error] .../linkerd/namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala:185: overloaded method value apply with alternatives:
[error]   (addrs: scala.collection.immutable.Set[com.twitter.finagle.Address],metadata: com.twitter.finagle.Addr.Metadata)com.twitter.finagle.Addr.Bound <and>
[error]   (addrs: com.twitter.finagle.Address*)com.twitter.finagle.Addr
[error]  cannot be applied to (scala.collection.immutable.Set[com.twitter.finagle.Address.Inet], com.twitter.finagle.Addr.Metadata)
[error]             addr() = Addr.Bound(addrs.toSet, convertMeta(boundMeta))
[error]                           ^
[error] one error found
[error] (namerd-iface-interpreter-thrift/compile:compileIncremental) Compilation failed
```

Comment (adleong) [inline namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala]: ah, I think scala just needs some help with the type unification.  Try:

```
addr() = Addr.Bound(addrs.toSet[Address], convertMeta(boundMeta))
```

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala [hunk 14]
```diff
@@ -2,16 +2,30 @@ package io.buoyant.consul
...
   import CatalogNamer._
 
+  private[this] lazy val domainFuture: Future[String] =
+    agentApi.self().flatMap { self =>
+      val domain = self.Config
+        .flatMap(_.Domain)
+        .map(_.stripPrefix(".").stripSuffix("."))
+      domain match {
+        case Some(d) => Future.value(d)
+        case None => Future.exception(???)
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Sorry, I meant that you would use some real exception here to indicate that the consul response was malformed, or didn't contain the right info, or whatever.

Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: I asked about that in a comment above. :)

I see 2 options:
1) Some concrete exception - in that case can you recommend one? :)
2) Rather than exception fallback to `consul` - the default value according to Consul docs.

What would you prefer?

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala [hunk 15]
```diff
@@ -2,16 +2,30 @@ package io.buoyant.consul
...
-  includeTag: Boolean = false
+  catalogApi: v1.CatalogApi,
+  agentApi: v1.AgentApi,
+  includeTag: Boolean = false,
+  setHost: Boolean = false
 ) extends Namer {
 
   import CatalogNamer._
 
+  private[this] lazy val domainFuture: Future[String] =
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Something to note: the domain request will be fired once when the catalog namer does `init`, and then never again.  This means if that request fails for any reason, the domain will never be populated.  Worse, anything flatMapped from the domain future (like the rest of the body of `init` will never be executed).  

Similarly, this also means that we will never pick up changes if the domain value in consul changes.

Does the consul self API support long polling?  Can we have the domain be populated by a Var.async instead?

@olix0r any thoughts on this?

Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Unfortunately it doesn't support long polling. Would you like a `Var.async` with some interval? If yes, what the interval would be ok? Something like `10.seconds`?

File: consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+  val client: Client,
+  val uriPrefix: String,
+  val backoffs: Stream[Duration] = Backoff.exponentialJittered(1.milliseconds, 5.seconds),
+  val stats: StatsReceiver = DefaultStatsReceiver
+) extends BaseApi with Closable {
+
+  val agentPrefix = s"$uriPrefix/agent"
+
+  // https://www.consul.io/docs/agent/http/agent.html#agent_self
+  def self(retry: Boolean = false): Future[LocalAgent] = {
```
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/v1/AgentApi.scala]: I'd rename this method to `localAgent` too

File: project/LinkerdBuild.scala [hunk 17]
```diff
@@ -12,7 +12,8 @@ object LinkerdBuild extends Base {
   val Bundle = config("bundle") extend Minimal
   val Dcos = config("dcos") extend Bundle
 
-  val consul = projectDir("consul")
+  lazy val consul = projectDir("consul")
```
Comment (adleong) [inline project/LinkerdBuild.scala]: We try to avoid lazy vals in our build file to keep dependancies explicit.  Adding a dependency on namer/core actually tangles things up a bit.  I actually think the best way to resolve this is to move the CatalogNamer out of `consul` and into `namer/consul`.

Comment (Ashald) [inline project/LinkerdBuild.scala]: OK, will do.

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala (hunk 14 above)
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Option 3) have domainFuture be a `Future[Option[String]]`.  When you use this value elsewhere, you can substitute a default value if it's None.

Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Thanks!

File: consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala (hunk 15 above)
Comment (adleong) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: 10s seems like a reasonable polling interval to me.  Just a warning that this is going to bit tricky to get right because `Var.async`s will stop updating if they don't have any listeners which will be true most of the time.  We probably don't want the `Var.async` to restart and fire a new request for every consul change.  (or maybe we do, in which case we can ditch `Var.async` and just make a Future call each time.)

Comment (Ashald) [inline consul/src/main/scala/io/buoyant/consul/CatalogNamer.scala]: Hm, sounds more complicated than I though first. In this case, would you mind making it 'just an another Future call each time' for now. Once we will get some usage experience with this we will be able to see the performance impact and if it will be significant then we will contribute another PR with an improvement.

What do you think?

File: namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala (hunk 13 above)
Comment (Ashald) [inline namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerClient.scala]: Thanks!

**Comment by Ashald** [comment]
@adleong I addressed you latest comments - can you take another look please?

File: namer/consul/src/main/scala/io/buoyant/namer/consul/CatalogNamer.scala [hunk 18]
```diff
@@ -108,30 +122,43 @@ class CatalogNamer(
       updateableAddr() = Addr.Pending
     }
 
-    def init(): Future[Unit] = mkRequest().flatMap(update).handle(handleUnexpected)
+    private[this] def metaFuture() =
```
Comment (adleong) [inline namer/consul/src/main/scala/io/buoyant/namer/consul/CatalogNamer.scala]: I don't think this refactor is necessary.  You should be able to just add a call to domainFuture in update (if setHost is set, of course).

File: namer/consul/src/main/scala/io/buoyant/namer/consul/CatalogNamer.scala [hunk 19]
```diff
@@ -108,30 +122,43 @@ class CatalogNamer(
...
+        }
+      else
+        Future.value(Addr.Metadata.empty)
+
+    def cycle(): Future[Unit] =
+      for {
+        services <- mkRequest()
+        meta <- metaFuture()
+        result <- update(services, meta)
+      } yield cycle().handle(handleUnexpected)
```
Comment (adleong) [inline namer/consul/src/main/scala/io/buoyant/namer/consul/CatalogNamer.scala]: This isn't quite right; `cycle` returns a Future so it shouldn't be in the yield position.

Comment (Ashald) [inline namer/consul/src/main/scala/io/buoyant/namer/consul/CatalogNamer.scala]: Can I do `yield None` and then a call to `cycle` on the next line? Or that's idiomatically incorrect?

File: namer/consul/src/main/scala/io/buoyant/namer/consul/CatalogNamer.scala (hunk 18 above)
Comment (Ashald) [inline namer/consul/src/main/scala/io/buoyant/namer/consul/CatalogNamer.scala]: So should I roll-back or this way it also OK?

Comment (adleong) [inline namer/consul/src/main/scala/io/buoyant/namer/consul/CatalogNamer.scala]: I suggest rolling back the refactor and just adding the domain call in update.

File: consul/src/test/scala/io/buoyant/consul/v1/AgentApiTest.scala [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+  def stubService(buf: Buf) = Service.mk[Request, Response] { req =>
+    val rsp = Response()
+    rsp.setContentTypeJson()
+    rsp.content = buf
+    rsp.headerMap.set("X-Consul-Index", "4")
+    lastUri = req.uri
+    Future.value(rsp)
+  }
+
+  test("self returns Self instance") {
```
Comment (adleong) [inline consul/src/test/scala/io/buoyant/consul/v1/AgentApiTest.scala]: update test name to not reference `Self` object

Comment (Ashald) [inline consul/src/test/scala/io/buoyant/consul/v1/AgentApiTest.scala]: Done

**Comment by adleong** [comment]
:star: Thanks for being patient with all my requests!  This looks good to me!

**Comment by olix0r** [comment]
👍  looks great!

**Comment by Ashald** [comment]
Thanks for your review guys! And sorry that this PR took so much of your time. :/ I'll try to go with smaller PRs in future.
